README.md

```diff
@@ -152,69 +151,39 @@ Upon a call to `wasi_thread_spawn`, the WASI host must:
-
-* TIDs are managed and provided by host.
-
-* TID 0 is reserved. `wasi_thread_spawn` should not return this value.
-
-  * It's widely assumed in musl/wasi-libc.
-
-* The upper-most 3-bits of TID are always 0.
-  `wasi_thread_spawn` should not return values with these bits set.
-
-  * The most significant bit is the sign bit. As `wasi_thread_spawn` uses
-    signed integer and uses negative values to indicate errors, a TID needs
-    to be positive.
-
-  * The second bit need to be 0 in order to be compatible with the TID-based
-    locking implementation in musl/wasi-libc.
-
-  * The third bit need to be 0 in order to make an extra room for other
-    reserved values in wasi-libc.
-    For example, it can be used to indicate the main thread, which doesn't
-    have a TID in the current version of this proposal.
-
-### Process
-
-* A process is a group of threads.
-
-* The main thread starts with a process which only contains
-  the main thread.
-
-* Threads created by a thread in a process using `wasi_thread_spawn`
-  are added to the process.
-
-* When a thread is terminated, it's removed from the process.
-
-### Voluntary thread termination
-
-A thread can terminate itself voluntarily by returning from
-`wasi_thread_start`.
-
-### Changes to WASI `proc_exit`
-
-With this proposal, the `proc_exit` function takes extra responsibility
-to terminate all threads in the process, not only the calling one.
-
-Any of the threads in the process can call `proc_exit`.
-
-### Traps
-
-When a trap occurs in any thread, the entire process is terminated.
-
-### Process exit status
-
-If one or more threads call WASI `proc_exit` or raise a trap,
-one of them is chosen by the runtime to represent the exit status
-of the process.
-It's non deterministic which one is chosen.
-
-If all the threads in the process have been terminated without calling
-`proc_exit` or raising a trap, it's treated as if the last thread called
-`proc_exit` with exit code 0.
+#### Design choice: thread IDs
+
+When `wasi_thread_spawn` successfully spawns a thread, it returns a thread ID
+(TID) &mdash; 32-bit integer with several restrictions. TIDs are managed and
+provided by the WASI host. To avoid compatibility issues in wasi-libc, `0` is
+reserved; `wasi_thread_spawn` should not return it. To avoid leaking
+information, the host may choose to return random TIDs (as opposed to leaking OS
+TIDs).
+
+TIDs returned by `wasi_thread_spawn` must always zero the uppermost three bits:
+- the most significant bit is the sign bit; `wasi_thread_spawn` uses negative
+  values to indicate errors, so a valid TID must zero this bit (i.e., a valid
+  TID is always positive)
+- the second most significant bit is used for TID-based locking in wasi-libc
+  &mdash; this must be zeroed initially in a valid TID
+- the third most significant bit allows wasi-libc to mark threads as the main
+  thread; it is zeroed by any thread returned by `wasi_thread_spawn`
```

Comment:
It seems odd to me to mention wasi-libc at all here.   Perhaps just say that the the valid  TID range is 1 - 2^29?

Remind me again why zero is reserved but also the third most significant bit?    Why would we even want more than one main thread?  Can't we just reserve a single value for that?

What is the TID-based locking thing again?

Reply:
i agree it's better to separate the spec (the tid range) and its rationales.

> Remind me again why zero is reserved but also the third most significant bit? Why would we even want more than one main thread? Can't we just reserve a single value for that?

0 is for the existing assumptions in musl.

the third bit is for the main thread. you are right a single value is enough for this purpose. however, as there are plenty of TID values and i opted to reserve more than absolute needs known right now.
reserving more values later would need to update every implementations. the opposite is easier.

> What is the TID-based locking thing again?

this kind of things:
https://github.com/WebAssembly/wasi-libc/blob/main/libc-top-half/musl/src/stdio/__lockfile.c

Reply:
Yeah, I did not actually want to explain all the details of what each bit meant; just kept it in because @yamt added it. These details felt very wasi-libc specific and ideally they would be explained in that repository. I would prefer to say that we reserve the value 0 and the top three bits (so 1 - 2^29 as @sbc100 points out) and vaguely say something about compatibility with other language implementations. Are we ok with that kind of change?

Reply:
That sounds good to me.

Reply:
i added reasons of the reservation because it was asked by @loganek . as he seems ok, i'm ok too.

Reply:
Ok, I modified the text to be less specific about the reasons "why" and added the $[1, 2^{29})$ range comment.